### PR TITLE
Update will cleanup unneeded artifacts.

### LINKED ISSUE #693: After an upgrade fails to download, subsequent upgrades will fail
We have reports that after an upgrade fails to download, subsequent upgrade attempts can fail with an error about a 0 byte file when trying to validate the sha and/or signature of the download.

More details needed cc @WiegerElastic

### LINKED ISSUE #694: Successful upgrades should cleanup old downloads
After an upgrade is successful, the downloaded files are not deleted in `Elastic/Agent/data/elastic-agent-*/downloads`. This can lead to several GB of unneeded data being preserved on disk after several subsequent upgrades.

We should delete the downloads files to save space once an upgrade has successfully completed.

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

The update process will cleanup unneeded artifacts. When an update
starts all artifacts that do not have the current version number in it's
name will be removed. If artifact retrieval fails, downloaded artifacts
are removed. On a successful upgrade, all contents of the downloads dir
will be removed.

## Why is it important?

If verification fails because of an issue with the download, the agent will not be able to upgrade future attempts.
If the agent goes through a lot of upgrades it will use a lot of storage for binaries that are not used.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

1. cherry-pick to `8.3`
2. Install agent running 8.4.0-SNAPSHOT
3. Upgrade agent through UI
4. After upgrade is successful, there is no `downloads` dir in old agent dir

## Related issues

- Closes #693 
- Closes #694

### CODE REVIEW

**Comment by joshdover** [comment]
@michel-laterman Will this cleanup logic apply only when a new upgrade action is created by the user, or is there internal retry logic that this will get used on as well? I think the behavior we want to have here is:
1. User initiates upgrade from UI
2. Agent receives the upgrade action from Fleet Server
3. Agent attempts to download/unpack upgrade tarball which may fail
4. Agent automatically attempts the whole upgrade again, cleaning up the failed upgrade attempt first. This time the download/unpack succeeds and the upgrade is completed.

In other words, users shouldn't need to retry the upgrade manually from the UI and agent should be able to handle this type of temporary error condition automatically.

**Comment by michel-laterman** [comment]
@joshdover,

I don't see any retry logic around the upgrader, we should create another issue (or PR) to implement it.
I think that extending the action-queue might be our best bet here as we would need to support
- retrying a scheduled action in the given window
- retrying a unscheduled action with some default timeout (1h for example)
If we were to add retry logic, the upgrader would need to change how the `preUpgradeCleanup` function I've added works so that the agent does not redownload successfully downloaded artifacts if the issue occurred during installation.

**Comment by joshdover** [comment]
> I don't see any retry logic around the upgrader, we should create another issue (or PR) to implement it.

@michel-laterman Thanks for looking into it. Let's open a new issue with these details, I think this may be something we want to prioritize and discuss before we start work on it.

File: internal/pkg/agent/application/upgrade/cleanup_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	err := preUpgradeCleanup("8.4.0")
+	require.NoError(t, err)
+
+	files, err := os.ReadDir(paths.Downloads())
+	require.NoError(t, err)
+	require.Len(t, files, 1)
+	require.Equal(t, "test-8.4.0-file", files[0].Name())
+	fi, err := files[0].Info()
+	require.NoError(t, err)
+	require.Greater(t, fi.Size(), int64(0))
```
Comment (AndersonQ) [inline internal/pkg/agent/application/upgrade/cleanup_test.go:51]: are you just checking that the file isn't empty? Wouldn't be more precise to test for the actual content of the file? They're static anyway.

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 2]
```diff
@@ -137,6 +142,11 @@ func (u *Upgrader) Upgrade(ctx context.Context, a Action, reexecNow bool) (_ ree
 	sourceURI := u.sourceURI(a.SourceURI())
 	archivePath, err := u.downloadArtifact(ctx, a.Version(), sourceURI)
 	if err != nil {
+		// Run the same preUpgradeCleanup task to get rid of any newly downloaded files
+		// This may have an issue if users are upgrading to the same version number.
+		if dErr := preUpgradeCleanup(a.Version()); dErr != nil {
+			u.log.Errorf("Unable to remove file after verification failure: %v", dErr)
+		}
 		return nil, err
```
Comment (AndersonQ) [inline internal/pkg/agent/application/upgrade/upgrade.go:150]: [Blocker]
 - Is it correct, the test is for `dErr`, but `err` is returned?
 - How would be possible to update to the same version? Anyway, if that is the case, the clean up is ineffective. Wouldn't be better just to remove all files on the folder? As I suggested above, we could have just the `cleanAllDownloads` and get rid of everything on every upgrade.

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 3]
```diff
@@ -184,6 +194,13 @@ func (u *Upgrader) Upgrade(ctx context.Context, a Action, reexecNow bool) (_ ree
 		return nil, nil
 	}
 
+	// Clean everything from the downloads dir
+	// If we wanted to be a bit simpler we could call os.RemoveAll to remove the dir + children
```
Comment (AndersonQ) [inline internal/pkg/agent/application/upgrade/upgrade.go:198]: That sounds like a plan :)

**Review by AndersonQ — CHANGES_REQUESTED** [review]
There is just one thing I believe is a blocker, besides that just questions and suggestions

File: internal/pkg/agent/application/upgrade/cleanup.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+			if err := os.Remove(filepath.Join(paths.Downloads(), file.Name())); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// cleanAllDownloads will remove all files from the downloads directory
+func cleanAllDownloads() error {
```
[the remark this answers is not in the document]
Comment (michel-laterman) [reply internal/pkg/agent/application/upgrade/cleanup.go:35]: No, we want to keep all artifacts with the current version number in case of a failure/rollback.

File: internal/pkg/agent/application/upgrade/cleanup.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	if err != nil {
+		return err
+	}
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		if err := os.Remove(filepath.Join(paths.Downloads(), file.Name())); err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (michel-laterman) [reply internal/pkg/agent/application/upgrade/cleanup.go]: i've used multierror, i think having the caller control errors makes it clearer when an error occurs

File: internal/pkg/agent/application/upgrade/upgrade.go (hunk 2 above)
Comment (michel-laterman) [reply internal/pkg/agent/application/upgrade/upgrade.go:150]: yes, we don't want to fail the upgrade because an artifact can't be deleted
the artifacts in downloads are used if a new binary is added as an input

File: internal/pkg/agent/application/upgrade/cleanup_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+	err := os.WriteFile(filepath.Join(dir, "test-8.3.0-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+	err = os.WriteFile(filepath.Join(dir, "test-8.4.0-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+	err = os.WriteFile(filepath.Join(dir, "test-8.5.0-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+	err = os.WriteFile(filepath.Join(dir, "test-hash-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+}
```
Comment (ph) [inline internal/pkg/agent/application/upgrade/cleanup_test.go:30]: A bit late to the discussion here, but there is an `FS Interface` at https://pkg.go.dev/io/fs#FS, this allow you mock everything in memory and not have to deal with the weird syncing issues on drives.

**Review by ph — APPROVED** [review]
This look ok to me, I haven't tested it manually.

**Review by AndersonQ — APPROVED** [review]

File: internal/pkg/agent/application/upgrade/upgrade.go (hunk 3 above)
Comment (michel-laterman) [reply internal/pkg/agent/application/upgrade/upgrade.go]: done, also added the call before the exec now call

**Comment by michel-laterman** [comment]
@blakerouse, when we work on the downloader for v2 we'll have to keep these fixes
For 8.4 the `downloads` dir has artifacts for 8.4 (`metricbeat-8.4.0` for example)
